dohclient.go

```diff
@@ -275,66 +273,139 @@ func dohQuicTransport(endpoint string, opt DoHClientOptions) (http.RoundTripper,
-	resp, err := r.RoundTripper.RoundTrip(req)
+func (s *quicConnection) OpenStream() (quic.Stream, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	stream, err := s.EarlyConnection.OpenStream()
 	if netErr, ok := err.(net.Error); ok && (netErr.Timeout() || netErr.Temporary()) {
-		r.pool.closeAll()
-		r.RoundTripper.Close()
-		resp, err = r.RoundTripper.RoundTrip(req)
+		Log.WithError(err).Debug("temporary fail when trying to open stream, attempting new connection")
+		if err = quicRestart(s); err != nil {
+			return nil, err
+		}
+		stream, err = s.EarlyConnection.OpenStream()
 	}
-	return resp, err
+	return stream, err
 }
 
-// UDP connection pool. Also a workaround for for the http3.RoundTripper. When using a custom
-// dialer that open its own UDP connections, http3.RoundTripper doesn't close them when the
-// remote terminates a connection, or when calling Close(). So we have to keep track of the
-// connections and close them all before calling Close() on the http3.RoundTripper.
-type udpConnPool struct {
-	conns []*net.UDPConn
-	mu    sync.Mutex
+func (s *quicConnection) NextConnection() quic.Connection {
+	return nil
 }
 
-func (p *udpConnPool) add(conn *net.UDPConn) {
-	if p == nil {
-		return
-	}
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	p.conns = append(p.conns, conn)
+func quicRestart(s *quicConnection) error {
+	// Try to open a new connection, but clean up our mess before we do so
+	// This function should be called with the quicConnection locked, but lock checking isn't provided
+	// in golang; the issue was closed with "Won't fix"
+	_ = s.EarlyConnection.CloseWithError(DOQNoError, "")
+
+	// We need to close the UDP socket ourselves as we own the socket not the quic-go module
+	// c.f. https://github.com/quic-go/quic-go/issues/1457
+	if s.udpConn != nil {
+		_ = s.udpConn.Close()
+		s.udpConn = nil
+	}
+	Log.WithFields(logrus.Fields{
+		"protocol": "quic",
+		"hostname": s.hostname,
+		"local":    s.lAddr.String(),
+		"remote":   s.rAddr,
+	}).Debug("attempt reconnect")
+	var err error
+	var earlyConn quic.EarlyConnection
+	earlyConn, s.udpConn, err = quicDial(s.hostname, s.rAddr, s.lAddr, s.tlsConfig, s.config)
+	if err != nil || s.udpConn == nil {
+		Log.WithFields(logrus.Fields{
+			"protocol": "quic",
+			"address":  s.hostname,
+			"local":    s.lAddr.String(),
+		}).WithError(err).Error("couldn't restart quic connection")
+		return err
+	}
+	Log.WithFields(logrus.Fields{
+		"protocol": "quic",
+		"address":  s.hostname,
+		"local":    s.lAddr.String(),
+		"rAddr":    s.rAddr,
+	}).Debug("restarted quic connection")
+
+	s.EarlyConnection = earlyConn
+	return nil
 }
 
-func (p *udpConnPool) closeAll() {
-	if p == nil {
-		return
+func quicDial(hostname, rAddr string, lAddr net.IP, tlsConfig *tls.Config, config *quic.Config) (quic.EarlyConnection, *net.UDPConn, error) {
+	udpAddr, err := net.ResolveUDPAddr("udp", rAddr)
+	if err != nil {
+		Log.WithError(err).Debug("couldn't resolve remote addr (" + rAddr + ") for UDP quic client")
```

Comment:
No need to change it now, but there's a Debugf() that may come in handy, same in the log below

Reply:
Or you could use
```go
Log.WithError(err).WithField("remote", rAddr).Debug("couldn't resolve address")
```

Reply:
Yeah, you're right. I'll send through a clean up patch later. Thanks.